File: src/typescript/domain/message/channel/voice/note-on.ts
import { ChannelVoiceMessage } from "../../interface";

/**
 * ノートオンメッセージを表すクラス。
 */
export default class NoteOnMessage implements ChannelVoiceMessage {
  /**
   * ノートナンバー。音程のこと。
   * 0を最も低い音程とする。最高は127。
   * デフォルトは60。中央のドを表す
   */
  noteNumber: number;

  /**
   * ノートオンを開始する、シーケンサー上のタイミング。
   */
  startedOn: number;

  /**
   * 発声時間。
   * ノートオンからノートオフまでの間とする。
   */
  duration: number;
  
  constructor(noteNumber: number = 60, startOn: number = 0, duration: number = 0){
    this.noteNumber = noteNumber;
    this.startedOn = startOn;
    this.duration = duration;
  }
}

File: src/python/domain/message/channel/voice/note_on.py
from mido import Message
from domain.message.interface import ChannelVoiceMessage
from domain.message.channel.voice.note_off import NoteOffMessage


class NoteOnMessage(ChannelVoiceMessage):
    def __init__(self, note_number: int, started_on: int, duration: int):
        self._note_number: int = note_number
        self._started_on: int = started_on
        self._duration: int = duration

    def toMIDIMessage(self, seek_time: int):
        # 直前のメッセージからの待機時間。シーケンス上の開始位置から、経過時間(seek_time)を引くことで算出する
        time = self._started_on - seek_time
        return Message("note_on", note=self._note_number, time=time)

    def toNoteOffMessage(self):
        # ノートオフメッセージの開始位置は、ノートオンの開始位置＋発声時間
        note_of_at = self._started_on + self._duration
        return NoteOffMessage(self._note_number, started_on=note_of_at)
